src/sinol_make/commands/run/__init__.py

```diff
@@ -194,63 +199,134 @@ def compile(self, solution):
-			result["Status"] = "TL"
-		elif "Memory" in result.keys() and result["Memory"] > memory_limit:
-			result["Status"] = "ML"
-		elif "Status" not in result.keys():
-			result["Status"] = "RE"
-		elif result["Status"] == "OK":
-			if result["Time"] > time_limit:
-				result["Status"] = "TL"
-			elif result["Memory"] > memory_limit:
-				result["Status"] = "ML"
+					if key == "Time":
+						result.Time = self.parse_time(value)
+					elif key == "Memory":
+						result.Memory = self.parse_memory(value)
+					else:
+						setattr(result, key, value)
+
+		if timeout_exit_code == 35072:
+			result.Status = "TL"
+		elif getattr(result, "Time") is not None and result.Time > time_limit:
+			result.Status = "TL"
+		elif getattr(result, "Memory") is not None and result.Memory > memory_limit:
+			result.Status = "ML"
+		elif getattr(result, "Status") is None:
+			result.Status = "RE"
+		elif result.Status == "OK":
+			if result.Time > time_limit:
+				result.Status = "TL"
+			elif result.Memory > memory_limit:
+				result.Status = "ML"
 			elif os.system("diff -q -Z %s %s >/dev/null"
-						% (output_file, self.get_output_file(test))):
-				result["Status"] = "WA"
+						% (output_file, answer_file)):
+				result.Status = "WA"
 		else:
-			result["Status"] = result["Status"][:2]
+			result.Status = result.Status[:2]
+
 		return result
 
-	def perform_executions(self, compiled_commands, names, solutions, report_file):
+
+	def execute_time(self, command, result_file, output_file, answer_file, time_limit, memory_limit):
+		timeout_exit_code = os.system(command)
+
+		result = ExecutionResult(None, None, None)
+		lines = open(result_file).readlines()
+		program_exit_code = None
+		if len(lines) == 3:
+			"""
+			If programs runs successfully, the output looks like this:
+			 - first line is CPU time in seconds
+			 - second line is memory in KB
+			 - third line is exit code
+			This format is defined by -f flag in time command.
+			"""
+			result.Time = round(float(lines[0].strip()) * 1000)
+			result.Memory = int(lines[1].strip())
+			program_exit_code = int(lines[2].strip())
+		if len(lines) > 0 and "Command terminated by signal " in lines[0]:
+			"""
+			If there was a runtime error, the first line is the error message with signal number.
+			For example:
+				Command terminated by signal 11
+			"""
+			program_exit_code = int(lines[0].strip().split(" ")[-1])
+
+		if program_exit_code != None and program_exit_code != 0:
+			result.Status = "RE"
+		elif timeout_exit_code != 0:
+			result.Status = "TL"
+		elif result.Time > time_limit:
+			result.Status = "TL"
+		elif result.Memory > memory_limit:
+			result.Status = "ML"
+		elif os.system("diff -q -Z %s %s >/dev/null" % (output_file, answer_file)):
+			result.Status = "WA"
+		else:
+			result.Status = "OK"
+
+		return result
+
+
+	def run_execution(self, execution):
+		"""
+		Run an execution and return the result as ExecutionResult object.
+		"""
```

Comment:
Still, what is the `execution` argument? The function name suggests that it's somehow "running" the argument, but the first line of code suggest that this argument contains the results of a previously-done execution, but then, what does this function do?

Reply:
This function runs a solution on the specified test and returns status code and time and memory used. I don't really have an idea how to call this function

Reply:
Maybe `def run_solution_on_test(test_info)` ?

Reply:
I would say that `run_solution(single_test_info)` is enough -- it avoids a repetition and underlines that it's a single test (though I'm not sure if it's actually the case -- I'm making the assumption based on the name `test_info` instead of `tests_info`). Though then, what kind of solution does it run, e.g. how does it know which solution to run? I guess it's also in the `test_info`, but that means that the name is a bit misleading.

Reply:
Then maybe `execution_info` or `run_info`? But then again, it uses the unhelpful `execution` word.

Reply:
`data_for_execution` implies that it contains multiple details about the not-yet-done execution that will be performed in this function, which I think is a quite readable name. To even more improve the readability, we can consider creating a class that will contain fields for `data_for_execution`, and then passing this class as an argument.

Reply:
I added the struct for execution data 4eb4acbc0273b501d174e5eed6bf132288a26a0b